File: Python/104_Max_depth_BT.py
from typing import Optional
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def solve(self, root):
        if root is None:
            return 0

        left = self.solve(root.left)
        right = self.solve(root.right)

        return 1 + max(left, right)

    def maxDepth(self, root: Optional[TreeNode]) -> int:
        return self.solve(root)


# -------------------------
# TAKE INPUT FROM USER
# -------------------------

values = input("Enter tree: ").split()


# -------------------------
# CREATE TREE
# -------------------------

if values[0].lower() == "null":
    root = None
else:
    root = TreeNode(int(values[0]))

    queue = deque([root])
    i = 1

    while queue and i < len(values):

        current = queue.popleft()

        # Left child
        if i < len(values) and values[i].lower() != "null":
            current.left = TreeNode(int(values[i]))
            queue.append(current.left)

        i += 1

        # Right child
        if i < len(values) and values[i].lower() != "null":
            current.right = TreeNode(int(values[i]))
            queue.append(current.right)

        i += 1


# -------------------------
# FIND MAXIMUM DEPTH
# -------------------------

solution = Solution()

answer = solution.maxDepth(root)

print("Maximum Depth:", answer)

File: CPP/TREE/104_Max_depth_BT.cpp
#include<iostream>
#include<algorithm>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x):val(x),left(nullptr),right(nullptr){}
};

int solve(TreeNode *root){
    if(root==nullptr) return 0;

    int left=solve(root->left);
    int right=solve(root->right);

    return 1+max(left,right);
}

int maxDepth(TreeNode *root){
    return solve(root);
}

int main(){
    // Create a simple binary tree for testing
    TreeNode *root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);

    int depth = maxDepth(root);
    cout << "Maximum depth of the binary tree: " << depth << endl;
}